[css-typed-om] support reification of transform functions

graouts opened the pull request:
#### 0b0b865747d48a813eee3c2f52da026714bd61aa
<pre>
[css-typed-om] support reification of transform functions
<a href="https://bugs.webkit.org/show_bug.cgi?id=245785">https://bugs.webkit.org/show_bug.cgi?id=245785</a>

Reviewed by Alex Christensen.

We add support for creating the concrete CSSTransformComponent classes from a
CSSFunctionValue passed to CSSStyleValueFactory::reifyValue(). This allows
the &quot;transform&quot; CSS property to be represented by the CSS Typed OM.

Creating those concrete classes is done by adding a new create(CSSFunctionValue&amp;)
method to the various CSSTransformComponent subclasses.

We also needed to add support for the primitive value type CSS_IDENT in order
to be able to handle a &quot;none&quot; keyword value for &quot;perspective(none)&quot;.

To pass tests, we also needed to implement CSSMatrixComponent::serialize().

There is still work to be done to handle CSSValueList correctly, so currently
only the first function provided to &quot;transform&quot; will be represented, but this
beyond the scope of this single bug.

* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/idlharness-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-normalization/normalize-ident.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-normalization/transformvalue-normalization.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-serialization/cssKeywordValue.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-serialization/cssTransformValue.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/computed/get-position-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/non-replaced-elements/form-controls/input-line-height-computed-expected.txt:
* Source/WebCore/css/typedom/CSSStyleValueFactory.cpp:
(WebCore::CSSStyleValueFactory::reifyValue):
* Source/WebCore/css/typedom/transform/CSSMatrixComponent.cpp:
(WebCore::CSSMatrixComponent::create):
(WebCore::CSSMatrixComponent::serialize const):
* Source/WebCore/css/typedom/transform/CSSMatrixComponent.h:
* Source/WebCore/css/typedom/transform/CSSPerspective.cpp:
(WebCore::CSSPerspective::create):
* Source/WebCore/css/typedom/transform/CSSPerspective.h:
* Source/WebCore/css/typedom/transform/CSSRotate.cpp:
(WebCore::CSSRotate::create):
* Source/WebCore/css/typedom/transform/CSSRotate.h:
* Source/WebCore/css/typedom/transform/CSSScale.cpp:
(WebCore::CSSScale::create):
* Source/WebCore/css/typedom/transform/CSSScale.h:
* Source/WebCore/css/typedom/transform/CSSSkew.cpp:
(WebCore::CSSSkew::create):
* Source/WebCore/css/typedom/transform/CSSSkew.h:
* Source/WebCore/css/typedom/transform/CSSSkewX.cpp:
(WebCore::CSSSkewX::create):
* Source/WebCore/css/typedom/transform/CSSSkewX.h:
* Source/WebCore/css/typedom/transform/CSSSkewY.cpp:
(WebCore::CSSSkewY::create):
* Source/WebCore/css/typedom/transform/CSSSkewY.h:
* Source/WebCore/css/typedom/transform/CSSTranslate.cpp:
(WebCore::CSSTranslate::create):
* Source/WebCore/css/typedom/transform/CSSTranslate.h:

Canonical link: <a href="https://commits.webkit.org/255022@main">https://commits.webkit.org/255022@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/76f66cce29208cafa4e50354d84fb27df5edd1a2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/91046 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/77/builds/31 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/21576 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/8/builds/100467 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/159337 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/95052 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/36 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/29112 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/83402 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/97171 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/96702 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/31 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/77824 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/27013 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/81987 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/21576 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/70047 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/35195 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/21576 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/67/builds/32992 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/21576 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/36774 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/77824 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/1536 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/38701 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/21576 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/css/typedom/transform/CSSMatrixComponent.cpp:
```diff
@@ -46,15 +50,103 @@ Ref<CSSTransformComponent> CSSMatrixComponent::create(Ref<DOMMatrixReadOnly>&& m
+                return Exception { TypeError, "Expected a CSSUnitValue."_s };
+            components.append(downcast<CSSUnitValue>(valueOrException.releaseReturnValue().get()).value());
+        }
+        return create(components);
```
This is an unnecessary vector copy.  Please WTFMove

graouts replied:
Addressed in 3bf495845111.

achristensen07 commented on Source/WebCore/css/typedom/CSSStyleValueFactory.cpp:
```diff
@@ -272,6 +283,50 @@ ExceptionOr<Ref<CSSStyleValue>> CSSStyleValueFactory::reifyValue(Ref<CSSValue> c
         
         return reifyValue(*valueList->begin(), document);
+    } else if (is<CSSFunctionValue>(cssValue)) {
+        auto makeTransformValue = [&]<class T>(ExceptionOr<Ref<T>> exceptionOrTransformComponent) -> ExceptionOr<Ref<CSSStyleValue>> {
```
does (auto& exceptionOrTransformComponent) work instead of the lambda template parameters?

graouts replied:
`auto&` didn't work but `auto` did. Addressed in f2c38a84e572.

achristensen07 commented on Source/WebCore/css/typedom/transform/CSSMatrixComponent.cpp:
```diff
@@ -46,15 +50,103 @@ Ref<CSSTransformComponent> CSSMatrixComponent::create(Ref<DOMMatrixReadOnly>&& m
+        });
+    case CSSValueMatrix3d:
+        return makeMatrix([](Vector<double> components) {
+            ASSERT(components.size() == 16);
```
I'm a little nervous about this being able to be hit then us reading out of bounds.  Could we add a size check, ASSERT_NOT_REACHED, and return an Exception of some kind?  Same with CSSValueMatrix above.

graouts replied:
Addressed in 6b8ab023c1ea.

achristensen07 commented on Source/WebCore/css/typedom/transform/CSSPerspective.cpp:
```diff
@@ -69,6 +73,27 @@ ExceptionOr<Ref<CSSPerspective>> CSSPerspective::create(CSSPerspectiveValue leng
+        return CSSPerspective::create("none"_s);
+    }
+
+    ASSERT(cssFunctionValue.size() == 1 && cssFunctionValue.item(0));
```
Let's make this a little more bullet proof with a check, ASSERT_NOT_REACHED, return an exception.

graouts replied:
Addressed in 6b8ab023c1ea.

achristensen07 commented on Source/WebCore/css/typedom/transform/CSSRotate.cpp:
```diff
@@ -70,6 +74,53 @@ ExceptionOr<Ref<CSSRotate>> CSSRotate::create(Ref<CSSNumericValue> angle)
+        });
+    case CSSValueRotate3d:
+        return makeRotate([](Vector<RefPtr<CSSNumericValue>> components) {
+            ASSERT(components.size() == 4);
```
Ditto here.  check components is the right size and the nullness is not unexpected, then assert_not_reached and return an exception if anything goes horribly wrong.

graouts replied:
Addressed in 6b8ab023c1ea.

achristensen07 commented on Source/WebCore/css/typedom/transform/CSSScale.cpp:
```diff
@@ -57,6 +61,53 @@ ExceptionOr<Ref<CSSScale>> CSSScale::create(CSSNumberish x, CSSNumberish y, std:
+                return Exception { TypeError, "Expected a CSSNumericValue."_s };
+            components.append(downcast<CSSNumericValue>(valueOrException.releaseReturnValue().ptr()));
+        }
+        return create(components);
```
WTFMove

graouts replied:
Addressed in 3bf495845111.

achristensen07 commented on Source/WebCore/css/typedom/transform/CSSScale.cpp:
```diff
@@ -57,6 +61,53 @@ ExceptionOr<Ref<CSSScale>> CSSScale::create(CSSNumberish x, CSSNumberish y, std:
 
+ExceptionOr<Ref<CSSScale>> CSSScale::create(CSSFunctionValue& cssFunctionValue)
+{
+    auto makeScale = [&](const Function<ExceptionOr<Ref<CSSScale>>(Vector<RefPtr<CSSNumericValue>>)>& create) -> ExceptionOr<Ref<CSSScale>> {
```
Can we make the parameter a Vector&& to prevent hard-to-see unnecessary vector copies?

graouts replied:
Addressed in 3bf495845111.

achristensen07 commented on Source/WebCore/css/typedom/transform/CSSMatrixComponent.cpp:
```diff
@@ -46,15 +50,103 @@ Ref<CSSTransformComponent> CSSMatrixComponent::create(Ref<DOMMatrixReadOnly>&& m
 
+ExceptionOr<Ref<CSSTransformComponent>> CSSMatrixComponent::create(CSSFunctionValue& cssFunctionValue)
+{
+    auto makeMatrix = [&](const Function<Ref<CSSTransformComponent>(Vector<double>)>& create) -> ExceptionOr<Ref<CSSTransformComponent>> {
```
Vector<double>&&

graouts replied:
Addressed in 3bf495845111.

achristensen07 commented on Source/WebCore/css/typedom/transform/CSSScale.cpp:
```diff
@@ -57,6 +61,53 @@ ExceptionOr<Ref<CSSScale>> CSSScale::create(CSSNumberish x, CSSNumberish y, std:
+        });
+    case CSSValueScale:
+        return makeScale([](Vector<RefPtr<CSSNumericValue>> components) {
+            ASSERT(components.size() == 1 || components.size() == 2);
```
please make more robust against future changes making these assertions wrong.

graouts replied:
Addressed in 6b8ab023c1ea.

achristensen07 commented on Source/WebCore/css/typedom/transform/CSSTranslate.cpp:
```diff
@@ -57,6 +60,55 @@ ExceptionOr<Ref<CSSTranslate>> CSSTranslate::create(Ref<CSSNumericValue> x, Ref<
+        });
+    case CSSValueTranslate3d:
+        return makeTranslate([](Vector<Ref<CSSNumericValue>> components) {
+            ASSERT(components.size() == 3);
```
check these, ASSERT_NOT_REACHED, return exception.

graouts replied:
Addressed in 6b8ab023c1ea.

achristensen07 commented on Source/WebCore/css/typedom/transform/CSSScale.cpp:
```diff
@@ -63,7 +63,7 @@ ExceptionOr<Ref<CSSScale>> CSSScale::create(CSSNumberish x, CSSNumberish y, std:
 ExceptionOr<Ref<CSSScale>> CSSScale::create(CSSFunctionValue& cssFunctionValue)
 {
-    auto makeScale = [&](const Function<ExceptionOr<Ref<CSSScale>>(Vector<RefPtr<CSSNumericValue>>&&)>& create) -> ExceptionOr<Ref<CSSScale>> {
+    auto makeScale = [&](const Function<ExceptionOr<Ref<CSSScale>>(Vector<RefPtr<CSSNumericValue>>&&)>& create, size_t minNumberOfComponents, std::optional<size_t> maxNumberOfComponents = std::nullopt) -> ExceptionOr<Ref<CSSScale>> {
```
Can this be size_t maxNumberOfComponents = minNumberOfComponents instead?  That would slightly simplify the code below.

graouts replied:
I had the same idea, but the compiler disagreed:

```
css/typedom/transform/CSSScale.cpp:66:174: error: default argument references parameter 'minNumberOfComponents'
    auto makeScale = [&](const Function<ExceptionOr<Ref<CSSScale>>(Vector<RefPtr<CSSNumericValue>>&&)>& create, size_t minNumberOfComponents, size_t maxNumberOfComponents = minNumberOfComponents) -> ExceptionOr<Ref<CSSScale>> {
```

achristensen07 commented on Source/WebCore/css/typedom/transform/CSSTranslate.cpp:
```diff
@@ -62,7 +62,7 @@ ExceptionOr<Ref<CSSTranslate>> CSSTranslate::create(Ref<CSSNumericValue> x, Ref<
 ExceptionOr<Ref<CSSTranslate>> CSSTranslate::create(CSSFunctionValue& cssFunctionValue)
 {
-    auto makeTranslate = [&](const Function<ExceptionOr<Ref<CSSTranslate>>(Vector<Ref<CSSNumericValue>>&&)>& create) -> ExceptionOr<Ref<CSSTranslate>> {
+    auto makeTranslate = [&](const Function<ExceptionOr<Ref<CSSTranslate>>(Vector<Ref<CSSNumericValue>>&&)>& create, size_t minNumberOfComponents, std::optional<size_t> maxNumberOfComponents = std::nullopt) -> ExceptionOr<Ref<CSSTranslate>> {
```
ditto

graouts committed:
[css-typed-om] support reification of transform functions
https://bugs.webkit.org/show_bug.cgi?id=245785

Reviewed by Alex Christensen.

We add support for creating the concrete CSSTransformComponent classes from a
CSSFunctionValue passed to CSSStyleValueFactory::reifyValue(). This allows
the "transform" CSS property to be represented by the CSS Typed OM.

Creating those concrete classes is done by adding a new create(CSSFunctionValue&)
method to the various CSSTransformComponent subclasses.

We also needed to add support for the primitive value type CSS_IDENT in order
to be able to handle a "none" keyword value for "perspective(none)".

To pass tests, we also needed to implement CSSMatrixComponent::serialize().

There is still work to be done to handle CSSValueList correctly, so currently
only the first function provided to "transform" will be represented, but this
beyond the scope of this single bug.

* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/idlharness-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-normalization/normalize-ident.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-normalization/transformvalue-normalization.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-serialization/cssKeywordValue.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-serialization/cssTransformValue.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/computed/get-position-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/non-replaced-elements/form-controls/input-line-height-computed-expected.txt:
* Source/WebCore/css/typedom/CSSStyleValueFactory.cpp:
(WebCore::CSSStyleValueFactory::reifyValue):
* Source/WebCore/css/typedom/transform/CSSMatrixComponent.cpp:
(WebCore::CSSMatrixComponent::create):
(WebCore::CSSMatrixComponent::serialize const):
* Source/WebCore/css/typedom/transform/CSSMatrixComponent.h:
* Source/WebCore/css/typedom/transform/CSSPerspective.cpp:
(WebCore::CSSPerspective::create):
* Source/WebCore/css/typedom/transform/CSSPerspective.h:
* Source/WebCore/css/typedom/transform/CSSRotate.cpp:
(WebCore::CSSRotate::create):
* Source/WebCore/css/typedom/transform/CSSRotate.h:
* Source/WebCore/css/typedom/transform/CSSScale.cpp:
(WebCore::CSSScale::create):
* Source/WebCore/css/typedom/transform/CSSScale.h:
* Source/WebCore/css/typedom/transform/CSSSkew.cpp:
(WebCore::CSSSkew::create):
* Source/WebCore/css/typedom/transform/CSSSkew.h:
* Source/WebCore/css/typedom/transform/CSSSkewX.cpp:
(WebCore::CSSSkewX::create):
* Source/WebCore/css/typedom/transform/CSSSkewX.h:
* Source/WebCore/css/typedom/transform/CSSSkewY.cpp:
(WebCore::CSSSkewY::create):
* Source/WebCore/css/typedom/transform/CSSSkewY.h:
* Source/WebCore/css/typedom/transform/CSSTranslate.cpp:
(WebCore::CSSTranslate::create):
* Source/WebCore/css/typedom/transform/CSSTranslate.h:

Canonical link: https://commits.webkit.org/255022@main
